streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java

```diff
@@ -84,26 +96,53 @@ public StreamThreadStateStoreProvider(final StreamThread streamThread,
     }
 
     @SuppressWarnings("unchecked")
-    private <T> T validateAndListStores(final StateStore store, final QueryableStoreType<T> queryableStoreType, final String storeName, final TaskId taskId) {
+    private static <T> T validateAndListStores(final StateStore store,
+                                               final QueryableStoreType<T> queryableStoreType,
+                                               final String storeName,
+                                               final TaskId taskId,
+                                               final boolean bypassCache) {
         if (store != null && queryableStoreType.accepts(store)) {
             if (!store.isOpen()) {
                 throw new InvalidStateStoreException(
                         "Cannot get state store " + storeName + " for task " + taskId +
                             " because the store is not open. " +
                             "The state store may have migrated to another instances.");
             }
+            final StateStore maybeUncached = maybeBypassCache(store, bypassCache);
             if (store instanceof TimestampedKeyValueStore && queryableStoreType instanceof QueryableStoreTypes.KeyValueStoreType) {
-                return (T) new ReadOnlyKeyValueStoreFacade<>((TimestampedKeyValueStore<Object, Object>) store);
+                return (T) new ReadOnlyKeyValueStoreFacade<>((TimestampedKeyValueStore<Object, Object>) maybeUncached);
             } else if (store instanceof TimestampedWindowStore && queryableStoreType instanceof QueryableStoreTypes.WindowStoreType) {
-                return (T) new ReadOnlyWindowStoreFacade<>((TimestampedWindowStore<Object, Object>) store);
+                return (T) new ReadOnlyWindowStoreFacade<>((TimestampedWindowStore<Object, Object>) maybeUncached);
             } else {
-                return (T) store;
+                return (T) maybeUncached;
             }
         } else {
             return null;
         }
     }
 
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    private static StateStore maybeBypassCache(final StateStore store, final boolean bypassCache) {
+        if (!bypassCache) {
+            return store;
+        } else if (store instanceof MeteredStore) {
+            final MeteredStore meteredStore = (MeteredStore) store;
+            final StateStore wrapped = meteredStore.wrapped();
+            final StateStore uncached = maybeBypassCache(wrapped, bypassCache);
+            if (uncached == wrapped) {
+                return store;
+            } else {
+                return meteredStore.reWrap(uncached);
```

Comment:
And this is where, if we did actually strip out the caching layer, we restore the layers "above" it.